cpp/src/io/text/multibyte_split.cu

```diff
@@ -0,0 +1,543 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/io/text/data_chunk_source.hpp>
+#include <cudf/io/text/multistate.hpp>
+#include <cudf/io/text/trie.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <rmm/cuda_stream_pool.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+
+#include <cub/block/block_load.cuh>
+#include <cub/block/block_scan.cuh>
+#include <cub/warp/warp_reduce.cuh>
+
+#include <iostream>
+#include <memory>
+
+namespace {
+
+template <typename Dividend, typename Divisor>
+inline constexpr auto ceil_div(Dividend dividend, Divisor divisor)
+{
+  return dividend / divisor + (dividend % divisor != 0);
+}
+
+using multistate = cudf::io::text::multistate;
+
+enum class scan_tile_status : uint8_t {
```

Comment:
Move `scan_tile_status`, `scan_tile_state_view`, `scan_tile_state`, and `scan_tile_state_callback` to a separate header, maybe add tests for it.